plotly/tools.py

```diff
@@ -1269,3 +1277,628 @@ def return_figure_from_figure_or_data(figure_or_data, validate_figure):
+        self.delta_x = self.x[1] - self.x[0]
+        self.delta_y = self.y[1] - self.y[0]
+        self.val_x = self.x
+        self.val_y = self.y
+        # Set up spacing
+        self.blank = np.zeros((self.density, self.density))
+        self.spacing_x = len(self.x)/float(self.density-1)
+        self.spacing_y = len(self.y)/float(self.density-1)
+        self.trajectories = []
+        # Rescale speed onto axes-coordinates
+        self.u = self.u/(self.x[-1]-self.x[0])
+        self.v = self.v/(self.y[-1]-self.y[0])
+        self.speed = np.sqrt(self.u*self.u + self.v*self.v)
+        # Rescale u and v for integrations.
+        self.u *= len(self.x)
+        self.v *= len(self.y)
+        self.validate()
+        self.st_x = []
+        self.st_y = []
+        self.get_streamlines()
+        streamline_x, streamline_y = self.sum_streamlines()
+        arrows_x, arrows_y = self.get_streamline_arrows()
+
+    def validate(self):
+        """
+        Validates that args and kwargs meet criteria,
+        specifically that scale and arrow_scale are positive
+        and that x, y, u, and v are the same length
+
+        :raises: (ImportError) If numpy is not imported.
+        :raises: (ValueError) If scale or arrow_scale is < 1.
+        :raises: (ValueError) If density or arrow_scale is < 1.
+        :raises: (PlotlyError) If x and y are not evenly spaced lists or
+            1D arrays.
+        :raises: (PlotlyError) If u and v are not the same shape.
+        """
+        if _numpy_imported is False:
+            raise ImportError("To use TraceFactory.create_streamline()"
+                              " please import numpy as np")
+        if self.arrow_scale <= 0:
+            raise ValueError("arrow_scale must be > 0")
+        if self.density <= 0:
+            raise ValueError("density must be > 0")
+        for index in range(len(self.x)-1):
+            if (self.x[index + 1]-self.x[index])-(self.x[1]-self.x[0]) > .0001:
+                raise exceptions.PlotlyError("x must be a 1 dimmensional"
+                                             "evenly spaced array")
+        for index in range(len(self.y)-1):
+            if (self.y[index + 1]-self.y[index])-(self.y[1]-self.y[0]) > .0001:
+                raise exceptions.PlotlyError("y must be a 1 dimmensional"
+                                             "evenly spaced array")
+        if self.u.shape != self.v.shape:
+                raise exceptions.PlotlyError("u and v should both be 2d arrays"
+                                             "with the same dimmensions")
+
+    def blank_pos(self, xi, yi):
+        """
+        Set up postitions for trajectories to be used with rk4 function.
+        """
+        return int((xi/self.spacing_x)+0.5), int((yi/self.spacing_y)+0.5)
+
+    def value_at(self, a, xi, yi):
+        """
+        Set up for RK4 function, based on Bokeh's streamline code
+        """
+        if type(xi) == np.ndarray:
```

Comment:
:cow2: `isinstance` instead of `type` --> `if isinstance(xi, np.ndarray):`

Also, is `value_at` required? This seems like it forces users to have `numpy` installed, is this OK/intentional?